### ISSUE TITLE
Potential idea for new API

### ISSUE THREAD

**Issue by jnsbck** [berenslab/ephyspy#7]
# idea sketch for new API

it would be nice if each feature was a self contained class somehow. This would have the advantage that all feature related metadata could be initialised and stored in the object. However, will potentially lead to a lot of overhead and instantiations of object. Possible solution would be to instantiate each object only once and then passing it to `SweepSetFeatureExtractor` which can call `__call__` to compute the feature. feature would take the current job of `get_[sweep/sweepset]_[feature_name]`. The diagnostic functionality could also be contained in the feature. 

``` python
from abc import ABC, abstractmethod

class EphysFeature(ABC):
    def __init__(self):
        pass

    @abstractmethod
    def __call__(self, sweep):
        pass

    @abstractmethod
    def plot(self, sweep):
        pass


class SweepNumAPs(EphysFeature):
    def __init__(self):
        super().__init__()

    def __call__(self, sweep):
        pass

    def plot(self, sweep):
        pass

# instantiated once 
{"num_ap": SweepNumAPs()}
```

**Comment by jnsbck** [comment]
This is being worked on in  [rewrite_object_oriented](https://github.com/berenslab/py_ephys/tree/rewrite_object_oriented).

**Comment by jnsbck** [comment]
fully implemented and merged into main with 28398e899c86b4a3cd41d75ee3ba51c0302ac4bb.
